File: app/client2/src/js/CRDTs/GCounter.js
class GCounter {
    constructor(counter) {
        this.counter = counter
    }

    static zero() {
        return new GCounter(new Map())
    }

    value() {
        return Object.values(this.counter).reduce((acc, currentValue) => acc + currentValue, 0);
    }

    inc(replica, value) {
        this.counter.set(replica, (this.counter.get(replica) ?? 0) + value)
    }

    static merge(a, b) {

        let keys = Array.from(a.counter.keys()).concat(Array.from(b.counter.keys()))
        const uniqueKeys = new Set(keys);

        let mergedCounter = new Map()
        for (const k of uniqueKeys) {
            mergedCounter.set(k,
                Math.max(
                    (a.counter.get(k) ?? 0),
                    (b.counter.get(k) ?? 0)
                )
            );
        }
        return new GCounter(mergedCounter);
    }

    toString() {
        const entries = Array.from(this.counter.entries()).map(([key, value]) => `"${key}": ${value}`);
        return `{ ${entries.join(', ')} }`;
    }

}

module.exports = GCounter

File: src/crdt/GCounter.py
import os
import sys

# Define module
current_path = os.path.dirname(__file__) + '/../..'
sys.path.append(current_path)


class GCounter:
    def __init__(self, counter):
        self.counter = counter

    @staticmethod
    def zero():
        return GCounter({})

    def value(self):
        return sum(self.counter.values(), 0)

    def inc(self, replica, value):
        self.counter[replica] = self.counter.get(replica, 0) + value

    @staticmethod
    def merge(a, b):
        merged_counter = {k: max(a.counter.get(k, 0), b.counter.get(k, 0)) for k in set(a.counter) | set(b.counter)}
        return GCounter(merged_counter)

    def __str__(self):
        return str(self.counter)
